rules/coding-style/src/Rector/Throw_/AnnotateThrowablesRector.php

```diff
@@ -93,59 +117,77 @@ public function throwException(int $code)
-        $identifiedThrownThrowables = $this->identifyThrownThrowables($throw);
+        if ($node instanceof Throw_) {
+            $this->analyzeStmtThrow($node);
+        }
 
-        foreach ($phpDocInfo->getThrowsTypes() as $throwsType) {
-            if (! $throwsType instanceof ObjectType) {
-                continue;
-            }
+        if ($node instanceof FuncCall) {
+            $this->analyzeStmtFuncCall($node);
+        }
 
-            $className = $throwsType instanceof ShortenedObjectType
-                ? $throwsType->getFullyQualifiedName()
-                : $throwsType->getClassName();
+        return count($this->throwablesToAnnotate) > 0;
+    }
 
-            if (! in_array($className, $identifiedThrownThrowables, true)) {
-                continue;
-            }
+    private function annotateThrowables(Node $node): void
+    {
+        $callee = $this->identifyCallee($node);
 
-            return true;
+        if ($callee === null) {
+            return;
         }
 
-        return false;
+        $phpDocInfo = $callee->getAttribute(AttributeKey::PHP_DOC_INFO);
+
+        foreach ($this->throwablesToAnnotate as $throwableToAnnotate) {
+            $docComment = $this->buildThrowsDocComment($throwableToAnnotate);
+            $phpDocInfo->addPhpDocTagNode($docComment);
+        }
     }
 
-    private function identifyThrownThrowables(Throw_ $throw): array
+    private function analyzeStmtThrow(Throw_ $throw): void
     {
-        if ($throw->expr instanceof New_) {
-            return [$this->getName($throw->expr->class)];
+        $foundThrownThrowables = [];
+
+        // throw new \Throwable
+        if ($throw->expr instanceof New_ && $this->getName($throw->expr->class) !== null) {
+            $foundThrownThrowables[] = $this->getName($throw->expr->class);
         }
 
         if ($throw->expr instanceof StaticCall) {
-            return $this->identifyThrownThrowablesInStaticCall($throw->expr);
+            $foundThrownThrowables = $this->identifyThrownThrowablesInStaticCall($throw->expr);
         }
 
         if ($throw->expr instanceof MethodCall) {
-            return $this->identifyThrownThrowablesInMethodCall($throw->expr);
+            $foundThrownThrowables = $this->identifyThrownThrowablesInMethodCall($throw->expr);
         }
 
-        return [];
+        $alreadyAnnotatedThrowables = $this->extractAlreadyAnnotatedThrowables($throw);
+
+        $this->compareThrowables($foundThrownThrowables, $alreadyAnnotatedThrowables);
+    }
+
+    private function analyzeStmtFuncCall(FuncCall $funcCall): void
+    {
+        $name = $funcCall->name;
+        if (! $name instanceof FullyQualified) {
+            return;
+        }
+        $foundThrownThrowables = $this->extractFunctionAnnotatedThrows($name);
+        $alreadyAnnotatedThrowables = $this->extractAlreadyAnnotatedThrowables($funcCall);
+        $this->compareThrowables($foundThrownThrowables, $alreadyAnnotatedThrowables);
```

Comment:
I don't understand what `compareThrowables` really means. 
Could you think of better name?

Reply:
I called it `diffThrowables()` and make it use `array_diff()` instead of the `foreach`: faster and more elegant.